### PR TITLE
Rename methods on error library and add `attach_provider`

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

This makes the naming on methods more consistent and adds a method to attach a provider without changing the current context.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/1201893074552404/1201481007343178/f) _(internal)_

## 🚫 Blocked by

- #599 

## 🔍 What does this change?

- Rename `wrap_err` to `attach_message`
- Rename `provide_context` to `change_context`
    - This implies a new trait `Context: Display + Debug` and changes to the internal `Frame` representation
- Add `attach_provider`:
    ```rust
    fn attach_provider<P: Provider + Display + Debug>(self, provider: P) -> Self;
   ```
- Make the provider API a nightly-only feature to make this crate to be used on stable as well. (We will switch to the upstream implementation of `Provider` as soon as it lands on `core`)
- Adjust `Report` so that `Report` will always have a `Context` associated (forbid `Provider` as context) and adjust the macros
- Greatly improve internal macro documentation
- Add an example to `Context` how is it used
- Small other drive-bys:
    -  implying `std` feature when `hooks` are enabled
    - Switch implementation location from `From::from` and `Report::from_error`

## 📜 Does this require a change to the docs?

The docs were adjusted

## 📹 Demo

Example used in documentation
```rust
use std::{fmt, fs, io};

use error::{Context, IntoReport, Result, ResultExt};

#[derive(Debug)]
pub enum ConfigError {
    ParseError,
}

impl fmt::Display for ConfigError {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        ...
    }
}

// `Error` is not implemented for `ConfigError` for any reason, so implement `Context` manually.
impl Context for ConfigError {}

pub fn read_file(path: &str) -> Result<String, io::Error> {
    // Creates a `Report` from `io::Error`, the current context is `io::Error`
    fs::read_to_string(path).report()
}

pub fn parse_config(path: &str) -> Result<Config, ConfigError> {
    // The return type of `parse_config` requires another context. By calling `change_context`
    // the context may be changed.
    read_file(path).change_context(ConfigError::ParseError)?;

    ...
}
```

### CODE REVIEW

File: packages/engine/lib/error/src/frame.rs [hunk 1]
```diff
@@ -33,10 +40,27 @@ use crate::{
 pub struct Frame {
     inner: ManuallyDrop<Box<FrameRepr>>,
     location: &'static Location<'static>,
-    source: Option<Box<Frame>>,
+    pub(crate) source: Option<Box<Frame>>,
```
Comment (TimDiekmann) [inline packages/engine/lib/error/src/frame.rs:43]: Required for `Frames` implementation

File: packages/engine/lib/error/src/error.rs [hunk 2]
```diff
@@ -1,27 +1,15 @@
...
-        };
-
-        Self::from_frame(
-            Frame::from_std(error, Location::caller(), None),
-            #[cfg(nightly)]
-            backtrace,
-            #[cfg(feature = "spantrace")]
-            Some(tracing_error::SpanTrace::capture()),
-        )
+        Self::from_error(error)
```
Comment (TimDiekmann) [inline packages/engine/lib/error/src/error.rs:13]: Switched this and `from_error` implementation location

File: packages/engine/lib/error/src/frame.rs [hunk 3]
```diff
@@ -33,10 +40,27 @@ use crate::{
...
 }
 
 impl Frame {
+    #[cfg(nightly)]
+    pub(crate) fn from_provider<P>(
```
Comment (TimDiekmann) [inline packages/engine/lib/error/src/frame.rs:48]: This is required for `attach_provider`

File: packages/engine/lib/error/src/frame.rs [hunk 4]
```diff
@@ -119,27 +145,14 @@ impl Frame {
...
-            // Fallback for `T: Error` as `Error` is wrapped inside of `ErrorRepr`
-            // TODO: Remove fallback when `Provider` is implemented for `Error` upstream
-            self.inner.downcast::<ErrorRepr<T>>().map(|addr| {
-                // SAFETY: Dereferencing is safe as T has the same lifetimes as Self and casting
-                //   `ErrorRepr<T>` to `T` is safe as `ErrorRepr` is `#[repr(transparent)]`
-                unsafe { addr.cast().as_ref() }
-            })
-        });
-
-        downcasted
```
Comment (TimDiekmann) [inline packages/engine/lib/error/src/frame.rs:139]: Logic moved to `FrameRepr`

File: packages/engine/lib/error/src/frame.rs [hunk 5]
```diff
@@ -185,31 +198,62 @@ impl Drop for Frame {
...
+        // Empty definition as a contextual message does not convey provider information
+    }
+}
+
+/// Wrapper around [`Context`].
+///
+/// As [`Context`] does not necessarily implement [`Provider`], an empty implementation is provided.
+/// If a [`Provider`] is required use it directly instead of [`Context`].
+#[repr(transparent)]
+struct ContextRepr<C>(C);
```
Comment (TimDiekmann) [inline packages/engine/lib/error/src/frame.rs:237]: Required as `Context` does not implement `Provider`, but `Unerased` requires it on nightly

File: packages/engine/lib/error/src/frame.rs [hunk 6]
```diff
@@ -223,23 +267,26 @@ impl<C: Message> Provider for MessageRepr<C> {
...
 }
 
 #[cfg(feature = "std")]
+impl<C: Context> Context for ErrorRepr<C> {}
+
+#[cfg(all(nightly, feature = "std"))]
 impl<E: Error> Provider for ErrorRepr<E> {
     fn provide<'a>(&'a self, demand: &mut Demand<'a>) {
-        #[cfg(all(nightly, feature = "std"))]
+        // ErrorProviderSpec::provide(self, demand);
```
Comment (TimDiekmann) [inline packages/engine/lib/error/src/frame.rs:289]: Leftover from trying specialization (it's working, will do it later)

```suggestion
```

File: packages/engine/lib/error/src/frame.rs [hunk 7]
```diff
@@ -307,21 +354,21 @@ struct FrameRepr<C = ()> {
     _context: C,
 }
 
-impl<C> FrameRepr<C>
+impl<T> FrameRepr<T>
```
Comment (TimDiekmann) [inline packages/engine/lib/error/src/frame.rs:357]: `T` is more intuitive than `C`

File: packages/engine/lib/error/src/lib.rs [hunk 8]
```diff
@@ -228,23 +224,69 @@ pub use self::{
...
+/// pub fn parse_config(path: &str) -> Result<Config, ConfigError> {
+///     // The return type of `parse_config` requires another context. By calling `change_context`
+///     // the context may be changed.
+///     read_file(path).change_context(ConfigError::ParseError)?;
 ///
-/// Note: This is currently defined as trait but will be changed to be a trait alias as soon as
-///   it's stabilized
-// TODO: change to `pub trait Context = ...`
-//   Tracking issue: https://github.com/rust-lang/rust/issues/41517
-pub trait Message: fmt::Display + fmt::Debug + Send + Sync + 'static {}
```
Comment (TimDiekmann) [inline packages/engine/lib/error/src/lib.rs:246]: Previously, we had two traits "aliases", now we have `Message`, which is an alias, and `Context`, which is not. As users eventually have to implement `Context`, having another trait, which doesn't need to be implemented, may be confusing. Thus, deciding to remove the alias.

File: packages/engine/lib/error/src/lib.rs [hunk 9]
```diff
@@ -254,41 +296,27 @@ pub(crate) mod test_helper {
...
-    pub struct ContextA(pub u32);
+    pub struct ContextA;
 
     impl fmt::Display for ContextA {
         fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
             fmt.write_str("Context A")
         }
     }
 
-    impl Provider for ContextA {
```
Comment (TimDiekmann) [inline packages/engine/lib/error/src/lib.rs:271]: Please shout if you find another `Provider` implementation 🙃

File: packages/engine/lib/error/Cargo.toml [hunk 10]
```diff
@@ -19,7 +19,7 @@ rustc_version = "0.2.3"
 [features]
 default = ["std", "hooks"]
 std = []
-hooks = ["dep:once_cell"]
+hooks = ["dep:once_cell", "std"]
```
Comment (TimDiekmann) [inline packages/engine/lib/error/Cargo.toml:22]: `once_cell::sync` (The thing we are using for setting the hooks) requires `std`, thus implying it here as well (compile error otherwise). Added this to the crate level docs as well.

File: packages/engine/lib/error/src/lib.rs [hunk 11]
```diff
@@ -228,23 +224,69 @@ pub use self::{
...
+/// impl fmt::Display for ConfigError {
+///     # #[allow(unused_variables)]
+///     fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+///         # const _: &str = stringify! {
+///         ...
+///         # }; Ok(())
+///     }
+/// }
+///
+/// // `Error` is not implemented for `ConfigError` for any reason, so implement `Context` manually.
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/lib.rs:260]: (probably messed up line length)

File: packages/engine/lib/error/src/lib.rs (hunk 8 above)
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/lib.rs:246]: I was going to ask, I think the reasoning makes sense although the trait expressions over the place are pretty unwieldy. We can always add it back if need be

File: packages/engine/lib/error/src/frame.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:43]: Fair call
```suggestion
    /// Required for [`Frames`] implementation for non-nightly builds
    ///
    /// [`Frames`]: crate::iter::Frames
    pub(crate) source: Option<Box<Frame>>,
```

Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/frame.rs:43]: Wait is this for the `source` field or the reasoning for `pub(crate)`?

File: packages/engine/lib/error/src/macros.rs [hunk 12]
```diff
@@ -1,62 +1,99 @@
...
+        //! This is a stable implementation for specialization (only possible within macros, as
+        //! there is no trait bound for this things).
+        //!
+        //! The different tags [`ReportTag`], [`ErrorTag`], and [`ContextTag`] have a blanket
+        //! implementation returning a concrete type. This type is then used to create a [`Report`].
+        //!
+        //! [`ErrorTag`] is implemented for `T: `[`Error`]s, [`ContextTag`] is implemented for `&T:
+        //! `[`Context`]s. As [`ContextTag`] is implemented references, [`ErrorTag`] has a higher
+        //! precedence and will be picked up first. So calling `my_error.kind()` will always
+        //! [`ErrorReporter`]. This can then be used to create a [`Report`] by calling
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/macros.rs:21]: Or are you describing the return type, so 'will always return'?
```suggestion
        //! will always return [`ErrorReporter`]. This can then be used to create a [`Report`] by calling
```

File: packages/engine/lib/error/src/report.rs [hunk 13]
```diff
@@ -19,27 +21,27 @@ use crate::{
 /// the [`Backtrace` documentation][`Backtrace`]. To enable the span trace, [`ErrorLayer`] has to
 /// be enabled.
 ///
-/// Context information can be added by using [`wrap()`] or [`ResultExt`]. The [`Frame`] stack can
-/// be iterated by using [`frames()`].
+/// Context information can be added by using [`attach_message()`] or [`ResultExt`]. The [`Frame`]
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/error/src/report.rs:24]: I think this phrasing is a bit confusing. We should find out how to phrase our (Message/Provider/Context) things better in plain English. Perhaps we just leave that for a final clean-up PR pre-release, can you make an Asana task (to tidy up the doc-strings for phrasing of (Message/Provider/Context)?

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]
This is **much** cleaner than it was before. I think we're reaching something a lot more intuitive (I hope).

Few small things and 🚀

File: packages/engine/lib/error/src/lib.rs [hunk 14]
```diff
@@ -165,19 +160,22 @@
...
-//! Feature     | Description                                                             | default
-//! ------------|-------------------------------------------------------------------------|---------
-//! `std`       | Enables support for [`Error`] and, on nightly, [`Backtrace`]            | enabled
-//! `hooks`     | Enables the usage of [`set_display_hook`] and [`set_debug_hook`]        | enabled
-//! `spantrace` | Enables the capturing of [`SpanTrace`]s                                 | disabled
-//! `futures`   | Provides a [`FutureExt`] adaptor                                        | disabled
+//!  Feature   | Description                                                    | implies | default
+//! -----------|----------------------------------------------------------------|---------|--------
+//!  `std`     | Enables support for [`Error`] and, on nightly, [`Backtrace`]   |         | enabled
+//!  `hooks`   |Enables the usage of [`set_display_hook`] and [`set_debug_hook`]| `std`   | enabled
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/lib.rs:166]: This would break `rustfmt` 🙃

[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/lib.rs:166]: Wait really? But all of the other ones have a space?...

File: packages/engine/lib/error/src/frame.rs (hunk 1 above)
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/frame.rs:43]: We could do this on a few other places nearby in the code if you'd like

File: packages/engine/lib/error/src/frame.rs [hunk 15]
```diff
@@ -267,9 +313,9 @@ impl VTable {
     /// # Safety
     ///
     /// - Layout of `Self` must match `FrameRepr<C>`.
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:315]: Generally, a `# Safety` section in a doc-string on an unsafe function means, that this is the contract you need to fulfill in order to call this function.

[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/frame.rs:315]: Sorry you're correct, I wasn't thinking clearly here (didn't properly notice this was a function's documentation)

File: packages/engine/lib/error/src/macros.rs [hunk 16]
```diff
@@ -1,62 +1,99 @@
 #[doc(hidden)]
 pub mod __private {
-    #![allow(clippy::unused_self)]
-    //! Uses [autoref-based stable specialization](https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md).
-    // TODO: Expand documentation when string literals are forbidden as this will shrink the
-    //   implementation by a fair bit.
+    //! Implementation detail for macros.
+    //!
+    //! ⚠️ **Functionality in this module is considered unstable and is subject to change at any time
+    //! without a major version bump!** ⚠️
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/macros.rs:6]: It's living inside of `pub mod __private`, so it *could* be used. Yes, the documentation will never be rendered and the name of the module is a good suggestion to not use it, but two lines don't hurt I guess. It's required to be public because the macros will use this module.

File: packages/engine/lib/error/src/report.rs (hunk 13 above)
Comment (TimDiekmann) [reply packages/engine/lib/error/src/report.rs:24]: [Asana task](https://app.asana.com/0/0/1202371933275761/f) (_internal_)

File: packages/engine/lib/error/src/lib.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/lib.rs:166]: It seems like they had a fix for it? https://github.com/rust-lang/rustfmt/issues/4210

File: packages/engine/lib/error/src/lib.rs [hunk 17]
```diff
@@ -228,23 +224,69 @@ pub use self::{
...
-/// Trait alias for a contextual message.
+/// Every [`Error`] also can act as a `Context`. If [`Error`] is not implemented or in a non-std
+/// environment, this trait can be implemented manually. In most cases, implementing this is not
+/// needed.
+///
+/// [`Error`]: std::error::Error
+///
+/// ## Example
+///
+/// Used for creating a [`Report`] or for switching the [`Report`]s context:
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/lib.rs:237]: We love English being ambiguous. In this instance it would be possessive, so the context belongs to the Report. 
https://www.thepunctuationguide.com/apostrophe.html

File: packages/engine/lib/error/src/macros.rs [hunk 18]
```diff
@@ -1,62 +1,99 @@
 #[doc(hidden)]
 pub mod __private {
-    #![allow(clippy::unused_self)]
-    //! Uses [autoref-based stable specialization](https://github.com/dtolnay/case-studies/blob/master/autoref-specialization/README.md).
-    // TODO: Expand documentation when string literals are forbidden as this will shrink the
-    //   implementation by a fair bit.
+    //! Implementation detail for macros.
+    //!
+    //! ⚠️ **Functionality in this module is considered unstable and is subject to change at any time
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/macros.rs:5]: I'm good with that reasoning, the symbols are fun

File: packages/engine/lib/error/src/macros.rs (hunk 16 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/macros.rs:6]: Ah that's unfortunate. Fair enough :)

File: packages/engine/lib/error/src/lib.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/lib.rs:166]: Somehow the fix doesn't seem to work. I changed the table layout to start and end with `'|'` but it's still converting back 😞

File: packages/engine/lib/error/src/frame.rs (hunk 1 above)
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:43]: I haven't found any code where this would be needed. This is the only location, where a field needs to be crate-visible, for functions this is only really helpful for `generalized` which already has reasoning on it.

File: packages/engine/lib/error/src/frame.rs [hunk 19]
```diff
@@ -333,15 +379,32 @@ where
...
+                return Some(addr.cast());
+            }
+
+            if let Some(addr) = (self.vtable.object_downcast)(self, TypeId::of::<ContextRepr<T>>())
+            {
+                return Some(addr.cast());
+            }
+
+            #[cfg(feature = "std")]
+            if let Some(addr) = (self.vtable.object_downcast)(self, TypeId::of::<ErrorRepr<T>>()) {
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:404]: Depends on how you added the `Error` to the report. When adding it by `change_context` or `from_context`, it will ignore the provider implementation until `Error` has a blanket implementation for `Provider`. Then it will be used directly.

File: packages/engine/lib/error/src/macros.rs (hunk 12 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/macros.rs:21]: The latter 😆

File: packages/engine/lib/error/src/frame.rs [hunk 20]
```diff
@@ -33,10 +40,27 @@ use crate::{
 pub struct Frame {
     inner: ManuallyDrop<Box<FrameRepr>>,
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:45]: I moved the logic to one function rather than having it at four functions. The SAFETY section were expanded: d7e784641c7b24d0da494aeab0ad37f2f686532d

File: packages/engine/lib/error/src/frame.rs [hunk 21]
```diff
@@ -185,31 +198,62 @@ impl Drop for Frame {
 /// are stored and called directly.
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:199]: Fixed the docstrings and also looked at other places for similar errors: 4e8790656f2071db3c98329c7c41d63ffb33066b

File: packages/engine/lib/error/src/frame.rs [hunk 22]
```diff
@@ -185,31 +198,62 @@ impl Drop for Frame {
...
-/// As [`Message`] does not necessarily implement [`Provider`], an empty implementation is provided.
-/// If a [`Provider`] is required use it directly instead of [`Message`].
-struct MessageRepr<M: Message>(M);
+/// A message does not necessarily implement [`Provider`], an empty implementation is provided.
+/// If a [`Provider`] is required attach it directly rather than attaching a message.
+struct MessageRepr<M>(M);
 
-impl<C: Message> fmt::Display for MessageRepr<C> {
-    fn fmt(&self, fmt: &mut Formatter<'_>) -> fmt::Result {
+impl<C: fmt::Display> fmt::Display for MessageRepr<C> {
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs]: Was fixed (here and at other locations) in f4879a2999235f96306a396cde648e2610e94bfa

File: packages/engine/lib/error/src/frame.rs [hunk 23]
```diff
@@ -307,21 +353,21 @@ struct FrameRepr<C = ()> {
...
-    C: Context,
+    T: Unerased,
 {
     /// Creates a new frame from a context
     ///
     /// # Safety
     ///
     /// Must not be dropped without calling `vtable.object_drop`
-    unsafe fn new(context: C) -> Box<FrameRepr> {
+    unsafe fn new(context: T) -> Box<FrameRepr> {
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs]: Fixed as drive-by in 4e8790656f2071db3c98329c7c41d63ffb33066b

File: packages/engine/lib/error/src/frame.rs [hunk 24]
```diff
@@ -16,6 +16,10 @@ use std::error::Error;
 use crate::provider::{self, Demand, Provider};
 use crate::Context;
 
+/// A [`Frame`] needs to implement all of the dependent traits so it's captured type needs these
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/error/src/frame.rs:19]:
```suggestion
/// A [`Frame`] needs to implement all of the dependent traits so its captured type needs these
```

File: packages/engine/lib/error/src/frame.rs [hunk 25]
```diff
@@ -5,18 +5,25 @@ use alloc::boxed::Box;
...
-use crate::{
-    provider::{self, Demand, Provider},
-    Context, Message,
-};
+#[cfg(nightly)]
+use crate::provider::{self, Demand, Provider};
+use crate::Context;
+
+#[cfg(nightly)]
+trait Unerased: Provider + fmt::Debug + fmt::Display + Send + Sync + 'static {}
```
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/frame.rs:24]: Better than it was 🎉 , I still think we might want to explain the naming, `Unerased`, a little more, maybe it's fine for it to only matter where it's used.

File: packages/engine/lib/error/src/frame.rs [hunk 26]
```diff
@@ -16,6 +16,10 @@ use std::error::Error;
 use crate::provider::{self, Demand, Provider};
 use crate::Context;
 
+/// A [`Frame`] needs to implement all of the dependent traits so it's captured type needs these
+/// traits as well.
+///
+/// This is a convenient trait alias to avoid writing down each trait explicitly every time.
```
Comment (Alfred-Mountfield) [inline packages/engine/lib/error/src/frame.rs:22]: Does this doc-comment appear on the other traits if under different cfgs, or do we have to copy paste it or something?

File: packages/engine/lib/error/src/frame.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/frame.rs:45]: I think I'd still kind of like to see a comment here saying why it's `ManuallyDrop` but if you're happy with it being on the actual `unsafe` portion then you can resolve this comment.

File: packages/engine/lib/error/src/frame.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/frame.rs:199]: Is it still accurate to say 
> to act on the associated context

If it's not specifically about a `Context` I think we should rephrase.

File: packages/engine/lib/error/src/lib.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/lib.rs:166]: Lets make an asana task to look into this as I don't think we should be struggling with phrasing and whitespaces just to not exceed a line-length.

How would we go about `skip`ping Rustfmt for this? Would it have to be the whole struct or could we skip it for just the doc-comment?

File: packages/engine/lib/error/src/frame.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:199]: Do you have a suggestion for rephrasing?

File: packages/engine/lib/error/src/frame.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:45]: The SAFETY section on an unsafe block should not explain why this is unsafe but why is it safe to call the function inside the unsafe block. The `FrameRepr` must not be dropped without using the vtable, and this is exactly why we use ManuallyDrop: We have a custom drop implementation that takes care of this.

File: packages/engine/lib/error/src/frame.rs (hunk 21 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/frame.rs:199]: Is it just the associated inner objects? The associated frames? Something else haha

File: packages/engine/lib/error/src/frame.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/frame.rs:45]: Which is why I'm not suggesting we necessarily add a "SAFETY" comment on the `inner` but it might be worth saying why we need `ManuallyDrop` at all on the `FrameRepr`. Was just saying we could add a comment exactly like you just wrote:
> The FrameRepr must not be dropped without using the vtable, so it's wrapped in ManuallyDrop, there's a custom drop implementation that takes care of this

Anyway the code somewhat speaks for itself, if someone wants to find the reasoning I'm sure they can infer it from finding `impl Drop`, I'm just saying we could save them the look

File: packages/engine/lib/error/src/lib.rs (hunk 14 above)
[the remark this answers is not in the document]
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/lib.rs:166]: Ah that's very frustrating. Fair enough. Let's still make an asana task to look into if we're doing something wrong or if we need to report the bug upstream.

File: packages/engine/lib/error/src/frame.rs [hunk 27]
```diff
@@ -297,7 +300,7 @@ impl VTable {
     ///
     /// # Safety
     ///
-    /// - Layout of `*frame` must match `FrameRepr<C>`.
+    /// - Layout of `*frame` must match `FrameRepr<T>`.
     unsafe fn object_drop<T>(frame: Box<FrameRepr>) {
         // Attach C's native vtable onto the pointer to `self._context`
```
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:305]: One yeah later: Hey, I found a C, lets change it to T 🙈

File: packages/engine/lib/error/src/frame.rs (hunk 26 above)
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/frame.rs:22]: Interesting, I didn't know that. I don't have the docs generated at the moment but does it show all definitions of the trait alias and the feature flags required for which one is active? Or does it just show the nightly one..

Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:22]: Oh, sorry, this might have been misleading: We are only using the nightly compiler because of the toolchain file to generate the docs, also docs.rs is using the nightly compiler. This does not apply generally

File: packages/engine/lib/error/src/frame.rs (hunk 25 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:25]: Described it even further 🙃

File: packages/engine/lib/error/src/frame.rs (hunk 20 above)
[the remark this answers is not in the document]
Comment (TimDiekmann) [reply packages/engine/lib/error/src/frame.rs:46]: Ah, now I get you 😄  Changed it again

File: packages/engine/lib/error/src/frame.rs (hunk 26 above)
Comment (Alfred-Mountfield) [reply packages/engine/lib/error/src/frame.rs]: Right but the error crate won't only be generated on nightly? I suspect people will want to be able to see nightly docs and stable docs? If that's not the case or isn't really convention then I don't mind.

Just want to know what happens if I run `cargo doc` on the error lib itself outside of hEngine. (We're also likely to move it out of the folder it's in soon aren't we?)

**Review by Alfred-Mountfield — APPROVED** [review]
